Source/Core/Core/Debugger/HLEFormatter.cpp

```diff
@@ -22,6 +29,232 @@ struct fmt::formatter<PowerPC::PairedSingle>
+    //   return std::find(values.begin(), values.end(), specifiers) != values.end();
+    // };
+
+    auto in_array = [&](auto... values) { return InArray(specifiers, values...); };
+
+    if (in_array("c"))  // Char values
+      return BaseType::Char;
+    else if (in_array("s"))
+      return BaseType::String;
+    else if (in_array("hhd", "hhi"))  // Signed values
+      return BaseType::Int8;
+    else if (in_array("hd", "hi"))
+      return BaseType::Int16;
+    else if (in_array("d", "i"))
+      return BaseType::Int32;
+    else if (in_array("ld", "li"))
+      return BaseType::Int64;
+    else if (in_array("hhb", "hhB", "hho", "hhO", "hhu", "hhx", "hhX"))  // Unsigned values
+      return BaseType::Uint8;
+    else if (in_array("hb", "hB", "ho", "hO", "hu", "hx", "hX"))
+      return BaseType::Uint16;
+    else if (in_array("b", "B", "o", "O", "u", "x", "X"))
+      return BaseType::Uint32;
+    else if (in_array("lb", "lB", "lo", "lO", "lu", "lx", "lX"))
+      return BaseType::Uint64;
+    else if (in_array("a", "A", "e", "E", "f", "F", "g", "G"))  // Floating-point values
+      return BaseType::Float;
+    else if (in_array("la", "lA", "le", "lE", "lf", "lF", "lg", "lG"))
+      return BaseType::Double;
+    else if (in_array("p"))  // Pointer value
+      return BaseType::Pointer;
+    return BaseType::Default;
+  }
+
+  void TranslateBaseFormat()
+  {
+    // Convert the type specifiers to the ones supported by {fmt}
+    for (auto c = base_format.begin(); c != base_format.end();)
+    {
+      if (*c == 'u')
+      {
+        *c = 'd';
+      }
+      else if (*c == 'h' || *c == 'l')
+      {
+        c = base_format.erase(c);
+        continue;
+      }
+      ++c;
+    }
+  }
+
+  template <typename ParseContext>
+  constexpr auto parse(ParseContext& ctx)
+  {
+    const auto begin = ctx.begin();
+    const auto end = ctx.end();
+    if (begin == end || *begin == '}')
+    {
+      base_format = "{}";
+      return begin;
+    }
+
+    // Let {fmt} handle the invalid end iterator
+    const auto error = begin;
+
+    // Search the closing brace
+    const auto format_end = std::find(begin, end, '}');
+    if (format_end == end)
+      return error;
+    base_format = "{:" + std::string(begin, format_end + 1);
+
+    // Get the format and type specifiers
+    const std::string_view type_view = ParseBaseType(begin, format_end);
+    base_type = GetBaseType(type_view);
+    if (base_type == BaseType::Default && !type_view.empty())
+      return error;
+
+    if (base_type == BaseType::Pointer)
+      base_format = "{:#010x}";
+
+    TranslateBaseFormat();
+
+    return format_end;
+  }
+
+  template <typename FormatContext>
+  auto format(const HLEArg& arg, FormatContext& ctx) const
+  {
+    const auto format_string = fmt::runtime(base_format);
+
+    switch (base_type)
+    {
+    default:
+    case BaseType::Default:
+    {
+      if (arg.IsFloat())
+      {
+        [[fallthrough]];
+      case BaseType::Double:
```

Comment:
I'm shocked this works.  Tbh, I don't find this very readable.

Is this just to avoid some duplication in something like this:

```
switch(base_type)
{
default:
case BaseType::Default:
{
  if (arg.IsFloat())
    return fmt::format_to(ctx.out(), format_string, arg.Value<double>());
  return fmt::format_to(ctx.out(), format_string, arg.Value<s64>());
}
case BaseType::Double:
  return fmt::format_to(ctx.out(), format_string, arg.Value<double>());
case BaseType::Int64:
  return fmt::format_to(ctx.out(), format_string, arg.Value<s64>());
...
}
```

Reply:
See https://en.wikipedia.org/wiki/Duff%27s_device for a related thing.

Reply:
@iwubcode 
Yeah, I used this syntax to avoid code duplication. You can think of `switch-case` like some kind of glorified `labels:`/`goto`. This technique is used in some memcpy implementation using Duff's device, there are also some concepts using it as well such as protothread (some kind of stackless/lightweight threads).

Reply:
I'm not a performance dev so I usually prefer readability unless something is a known hit.  This seems like premature optimization but I have no clue, maybe it could be a hotspot if this is ran often.  Regardless, I had not heard of this before, appreciate you both giving me details about it.  i will read some more.  Thanks!

Reply:
+1 from me for this being very weird and difficult to read.

Reply:
Alright, I'm going to replace it with a more readable syntax instead.